### ISSUE TITLE
Always throwing general exception even though i raise custom exception.

### ISSUE THREAD

**Issue by rbaddam09** [Azure/azure-functions-dotnet-worker#1042]
I implemented the custom middleware in azure function. Based on the different business cases, I'm throwing custom exception. I had switch case implemented based on the type of exception in the middle ware to catch the exception which is thrown from azure function and send the appropriate error message and the status code to the user. 
But somehow it always throwing the default exception and can not able to catch the custom exception and that is resulting into generic message to the user instead of user friendly message.

**Targeted Framework** : .net 6.0

![image](https://user-images.githubusercontent.com/65864404/190874511-a8f8ffae-255f-4d3d-b4df-f248c6ec4103.png)

**You can download the code from here:**
https://github.com/rbaddam09/Exception-Handling.git

**Middleware:**
```
public class ExceptionHandlingMiddleware : IFunctionsWorkerMiddleware
    {
        private readonly ILogger<ExceptionHandlingMiddleware> _logger;
        public ExceptionHandlingMiddleware(ILogger<ExceptionHandlingMiddleware> logger)
        {
            _logger = logger;
        }
        public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
        {
            try
            {
                await next(context);
            }
            catch (Exception exception)
            {
                var httpReqData = await context.GetHttpRequestDataAsync();
                AppError errorDesc;
                switch (exception)
                {
                    case AppException e:
                        errorDesc = new AppError()
                        {
                            Type = HttpStatusCode.NotFound.ToString(),
                            Detail = e.Message
                        };
                        var newHttpResponse1 = httpReqData.CreateResponse(HttpStatusCode.NotFound);
                        await newHttpResponse1.WriteAsJsonAsync(errorDesc, newHttpResponse1.StatusCode);
                        context.GetInvocationResult().Value = newHttpResponse1;
                        break;
                    default:
                        errorDesc = new AppError()
                        {
                            Type = HttpStatusCode.InternalServerError.ToString(),
                            Detail = "Something went wrong"
                        };
                        var newHttpResponse2 = httpReqData.CreateResponse(HttpStatusCode.NotFound);
                        await newHttpResponse2.WriteAsJsonAsync(errorDesc, newHttpResponse2.StatusCode);
                        context.GetInvocationResult().Value = newHttpResponse2;
                        break;
                }

            }
        }

    }
```
**Exception is thrown from azure function like below**
```
  public class Function1
    {
        private readonly ILogger _logger;

        public Function1(ILoggerFactory loggerFactory)
        {
            _logger = loggerFactory.CreateLogger<Function1>();
        }

        [Function("Function1")]
        public async Task<HttpResponseData> Run([HttpTrigger(AuthorizationLevel.Function,  "get" ,Route ="api/{env}/{id}" )] HttpRequestData req,string env, string id)
        {
            _logger.LogInformation("C# HTTP trigger function processed a request.");
            //if (string.IsNullOrEmpty(id))
            //{
                throw new AppException("name should not be empty");
            //}

            var response =  req.CreateResponse(HttpStatusCode.OK);
            response.Headers.Add("Content-Type", "text/plain; charset=utf-8");

            response.WriteString("Welcome to Azure Functions!");

            return response;
        }
    }
        
```
**Program.cs**
```
using FunctionApp2;
using Microsoft.Extensions.Hosting;

var host = new HostBuilder()
    .ConfigureFunctionsWorkerDefaults(
    builder =>
    {
        builder.UseMiddleware<ExceptionHandlingMiddleware>();
    }
    )
    .Build();

host.Run();
```

**Comment by kshyju** [comment]
@rbaddam09 Could you please format the code in the question? (see _[Creating and highlighting code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks)_) 

The code you shared is also not compliable as it is. Please fix that so that we can take a look. Thank you!

**Comment by rbaddam09** [comment]
@kshyju formatted the code and gave the git repo url where you can download and debug the code

**Comment by kshyju** [comment]
Thanks!

If your function is async and an exception is thrown while executing the function code, that exception is wrapped in an [AggregateException](https://learn.microsoft.com/en-us/dotnet/api/system.aggregateexception?view=net-6.0) before it is propagated.  You need to inspect the `InnerExceptions` property of this AggregateException instance and act on the inner exception entries as needed.

So in short, update your catch block to handle the AggregateException as well.

Hope this helps. Let us know if you still run into any issues.

**Comment by rbaddam09** [comment]
@kshyju thanks for looking into it.
I'm catching AggregateException . But i do not see anything which identify the type of custom exception . In that case i do not know what is the status code i should have to return to the user. Am i missing something here?
Not sure what you are referring here?. My title itself i mentioned that i can able to catch the general exception and read the message what is thrown. But the question here is that how would i know which status code i should have to return . What if the exception is thrown from un handled code and if i make use of the InnerExceptions message porperty i would end showing the message which thrown from the un handled code not the user friendly message.
![image](https://user-images.githubusercontent.com/65864404/190929958-5225d502-5fcb-4ac4-bd7b-67a9ee8d320b.png)

**Comment by kshyju** [comment]
@rbaddam09 The type of the inner exception entry will be the AppException type. 

Here is a working version of the Invoke method.

````
public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
{
    try
    {
        await next(context);
    }
    catch (Exception exception)
    {
        if (exception is AggregateException aggregateException)
        {
           // I chose to take the first from the InnerExceptions collection. You may check all of them as needed.
            exception = aggregateException.InnerExceptions.First();
        }
        _logger.LogError(exception, "Error processing invocation");

        switch (exception)
        {
            case AppException e:
                var response = await BuildErrorResponse(context, HttpStatusCode.NotFound, e.Message);
                context.GetInvocationResult().Value = response;
                break;
            default:
                response = await BuildErrorResponse(context, HttpStatusCode.InternalServerError, "Something wrong");
                context.GetInvocationResult().Value = response;
                break;
        }
    }
}

private async Task<HttpResponseData> BuildErrorResponse(FunctionContext context, HttpStatusCode statusCode, 
                                                                                                 string message)
{
    var errorDesc = new AppError()
    {
        Type = statusCode.ToString(),
        Detail = message
    };
    var httpReqData = await context.GetHttpRequestDataAsync();
    var newHttpResponse = httpReqData.CreateResponse(HttpStatusCode.NotFound);
    await newHttpResponse.WriteAsJsonAsync(errorDesc, newHttpResponse.StatusCode);

    return newHttpResponse;
}
````

![image](https://user-images.githubusercontent.com/144469/192855287-f90ee1b4-7d47-4aff-af42-1d46982ef8d1.png)

Hope this helps. Let us know if you still run into issues.
